File: apps/ios/BitOS/Features/Create/SharedTemplateStore.swift
import Foundation
import Observation

/// Relay-fetched shared templates (plan MST-045 / wave 3): subscribes to
/// kind-30078 app-data events, keeps the newest verified event per
/// template id whose d-tag addresses `com.bitos.bitz:template:*`, and
/// exposes summary rows via the tested bridge seam (raw tags+content ride
/// along so Apply re-parses through the shared contract — Android parity).
@MainActor
@Observable
final class SharedTemplateStore {
    struct Row: Identifiable {
        let id: String
        let label: String
        let emoji: String
        let priceSats: Int64
        let category: String
        let createdAt: Int64
        let tagsJson: String
        let content: String

        /// Bolt glyph comes from the Solar zap icon at the render site.
        var priceLabel: String? { priceSats > 0 ? "\(priceSats) sats" : nil }
    }

    private(set) var rows: [Row] = []

    private let pool: RelayPool
    private let client = FrameworkBusinessCoreClient()
    private var framesTask: Task<Void, Never>?
    private var requested = false

    init(pool: RelayPool) {
        self.pool = pool
    }

    /// Starts collecting frames (idempotent) and issues the REQ once.
    func start() {
        guard framesTask == nil else { return }
        Task { [weak self, pool, client] in
            let stream = await pool.frames()
            guard let self, !Task.isCancelled else { return }
            self.framesTask = FrameIngest.pump(
                stream: stream,
                isAlive: { [weak self] in self != nil },
                ingest: Self.templateIngest(client)
            ) { [weak self] row in
                await self?.upsert(row)
            }
        }
        guard !requested else { return }
        requested = true
        Task { [pool] in
            await pool.start()
            await pool.broadcast(
                #"["REQ","bitos-templates",{"kinds":[30078],"limit":200}]"#
            )
        }
    }

    /// Protocol gate + template extraction in one off-main step: decode,
    /// d-tag filter, summary bridge call and row construction all run on the
    /// ingest task; the main actor only upserts finished rows.
    private nonisolated static func templateIngest(
        _ client: FrameworkBusinessCoreClient
    ) -> @Sendable (RelayFrame) -> Row? {
        { frame in
            guard let event = client.decodeVerifiedEvent(message: frame.message, relay: frame.relay.rawValue),
                  event.kind == 30078,
                  let dTag = event.tags.first(where: { $0.first == "d" })?.dropFirst().first,
                  dTag.hasPrefix("com.bitos.bitz:template:") else { return nil }
            guard let tagsData = try? JSONSerialization.data(withJSONObject: event.tags),
                  let tagsJson = String(data: tagsData, encoding: .utf8) else { return nil }
            let summary = client.memeSharedTemplateSummary(tagsJson: tagsJson, content: event.content)
            guard !summary.isEmpty,
                  let data = summary.data(using: .utf8),
                  let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let id = root["id"] as? String, !id.isEmpty else { return nil }
            return Row(
                id: id,
                label: (root["label"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? id,
                emoji: (root["emoji"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? "🖼",
                priceSats: (root["priceSats"] as? NSNumber)?.int64Value ?? 0,
                category: (root["category"] as? String) ?? "meme",
                createdAt: event.createdAt,
                tagsJson: tagsJson,
                content: event.content
            )
        }
    }

    /// Newest-wins per template id; rail-capped (24), newest-first.
    private func upsert(_ row: Row) {
        if let existing = rows.first(where: { $0.id == row.id }), existing.createdAt >= row.createdAt {
            return
        }
        rows = (rows.filter { $0.id != row.id } + [row])
            .sorted { $0.createdAt > $1.createdAt }
            .prefix(24)
            .map { $0 }
    }
}

File: apps/android/src/main/kotlin/space/bitos/app/data/feed/SharedTemplateStore.kt
package space.bitos.app.data.feed

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import space.bitos.app.data.relay.RelayPool
import space.bitos.app.data.relay.VerifiedPoolFrame
import space.bitos.core.bridge.BusinessCoreBridge
import space.bitos.core.nostr.EventHasher
import space.bitos.core.nostr.NostrEventCodec
import space.bitos.core.nostr.Sha256EventHasher
import space.bitos.core.studio.MemeTemplateContract

/**
 * Relay-fetched shared templates (plan MST-045 / wave 3): subscribes to
 * kind-30078 app-data events, keeps the newest event per template id whose
 * d-tag addresses `com.bitos.bitz:template:*`, and exposes summary rows
 * (the raw tags+content ride along so Apply reuses the tested bridge
 * parse — never re-trusted client-side). Signature-verified; hostile
 * shapes drop inside the shared contract.
 */
class SharedTemplateStore(
    private val scope: CoroutineScope,
    private val pool: RelayPool,
    private val hasher: EventHasher = Sha256EventHasher,
) {
    data class Row(
        val id: String,
        val label: String,
        val emoji: String,
        val priceSats: Long,
        val category: String,
        val createdAt: Long,
        val tagsJson: String,
        val content: String,
    )

    private val bridge = BusinessCoreBridge()

    private val mutableRows = MutableStateFlow<List<Row>>(emptyList())
    val rows: StateFlow<List<Row>> = mutableRows.asStateFlow()

    private var requested = false
    private var collectJob: Job? = null

    init {
        collectJob = scope.launch {
            pool.verifiedFrames.collect { gated ->
                val event = (gated as? VerifiedPoolFrame.Verified)?.event ?: return@collect
                if (event.kind != SHARED_TEMPLATE_KIND) return@collect
                val dTag = event.tags.firstOrNull { it.firstOrNull() == "d" }?.getOrNull(1) ?: return@collect
                if (!MemeTemplateContract.isTemplateDTag(dTag)) return@collect
                val tagsJson = space.bitos.core.store.TagsCodec.encode(event.tags)
                val summary = bridge.memeSharedTemplateSummary(tagsJson, event.content)
                if (summary.isEmpty()) return@collect
                rowFrom(summary, event.createdAt, tagsJson, event.content)?.let(::upsert)
            }
        }
    }

    fun subscribe() {
        if (requested) return
        requested = true
        pool.broadcast(
            NostrEventCodec.encodeRequest(
                "bitos-templates",
                """{"kinds":[$SHARED_TEMPLATE_KIND],"limit":200}""",
            ),
        )
    }

    /** Newest-wins per template id; rail-capped, newest-first. */
    private fun upsert(row: Row) {
        val current = mutableRows.value
        val existing = current.firstOrNull { it.id == row.id }
        if (existing != null && existing.createdAt >= row.createdAt) return
        val next = (current.filterNot { it.id == row.id } + row)
            .sortedByDescending { it.createdAt }
            .take(MemeTemplateContract.MAX_SHARED_ROWS)
        mutableRows.value = next
    }

    private companion object {
        const val SHARED_TEMPLATE_KIND = 30_078
    }
}

/** Bridge summary row → store row (parses the tested JSON shape). */
private fun rowFrom(summaryJson: String, createdAt: Long, tagsJson: String, content: String): SharedTemplateStore.Row? =
    runCatching {
        val root = org.json.JSONObject(summaryJson)
        SharedTemplateStore.Row(
            id = root.optString("id"),
            label = root.optString("label").ifBlank { root.optString("id") },
            emoji = root.optString("emoji").ifBlank { "\uD83D\uDDBC" },
            priceSats = root.optLong("priceSats", 0L),
            category = root.optString("category").ifBlank { "meme" },
            createdAt = createdAt, tagsJson = tagsJson, content = content,
        )
    }.getOrNull()?.takeIf { it.id.isNotBlank() }
